cmd/clusterctl/pkg/client/config/providers_client.go

```diff
@@ -0,0 +1,203 @@
+type ProvidersClient interface {
+	// List returns all the provider configurations, including provider configurations hard-coded in clusterctl
+	// and user-defined provider configurations read from the clusterctl configuration file.
+	// In case of conflict, user-defined provider override the hard-coded configurations.
+	List() ([]Provider, error)
+
+	// Get return the configuration for the provider with a given name.
+	// In case the name does not correspond to any existing provider, an error is returned.
+	Get(name string) (*Provider, error)
+}
+
+// providersClient implements ProvidersClient.
+type providersClient struct {
+	reader Reader
+}
+
+// ensure providersClient implements ProvidersClient.
+var _ ProvidersClient = &providersClient{}
+
+func newProvidersClient(reader Reader) *providersClient {
+	return &providersClient{
+		reader: reader,
+	}
+}
+
+func (p *providersClient) defaults() []Provider {
+
+	// clusterctl includes an hard-coded list of "default" providers in order to provide to the users the simplest
+	// out-of-box experience. Other providers can be added by using the clusterctl configuration file.
+
+	// if you are a developer of a community maintained provider, you can send a PR to extend the following list.
+
+	defaults := []Provider{
+		// cluster API core provider
+		&provider{
+			name:         CAPI,
+			url:          "https://github.com/kubernetes-sigs/cluster-api/releases/latest/cluster-api-components.yaml",
+			providerType: clusterctlv1.CoreProviderType,
+		},
+
+		// Infrastructure providersClient
+		&provider{
+			name:         "aws",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases/latest/infrastructure-components.yaml",
+			providerType: clusterctlv1.InfrastructureProviderType,
+		},
+		&provider{
+			name:         "docker",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-provider-docker/releases/latest/provider_components.yaml",
+			providerType: clusterctlv1.InfrastructureProviderType,
+		},
+		&provider{
+			name:         "vsphere",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/releases/latest/infrastructure-components.yaml",
+			providerType: clusterctlv1.InfrastructureProviderType,
+		},
+
+		// Bootstrap providersClient
+		// TODO: CABPK in v1alpha3 will be included into CAPI, so this entry can be removed as soon as v1alpha3 is ready for test
+		&provider{
+			name:         "kubeadm",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/releases/latest/bootstrap-components.yaml",
+			providerType: clusterctlv1.BootstrapProviderType,
+		},
+	}
+
+	// ensure defaults are consistently sorted
+	sort.Slice(defaults, func(i, j int) bool {
+		return defaults[i].Name() < defaults[j].Name()
+	})
+
+	return defaults
+}
+
+func (p *providersClient) List() ([]Provider, error) {
+	// Creates a maps with all the defaults provider configurations
+	rMap := map[string]Provider{}
```

Comment:
Do we need this map because of the key? Given that the providers slice seems fixed in size (or small in general) these conversions might be overkill. If we create a ProviderList type as mentioned in the other comment, we can have a `Find()` or `GetByName` method that loops over the slice.

Reply:
Removed the map